File: rust-NN/op2.py
import numpy as np

class Array:
    def __init__(self, shape, data):
        self.shape = shape
        self.data = data

    def get_index(self, indices):
        flat_index = 0
        stride = 1
        for i in range(len(indices) - 1, -1, -1):
            flat_index += indices[i] * stride
            stride *= self.shape[i]
        return flat_index

    def get(self, indices):
        index = self.get_index(indices)
        return self.data[index]

    def broadcast_shape(self, other_shape):
        result_shape = []
        for s1, s2 in zip(reversed(self.shape), reversed(other_shape)):
            if s1 == s2 or s1 == 1 or s2 == 1:
                result_shape.append(max(s1, s2))
            else:
                raise ValueError("Incompatible shapes for broadcasting")
        return tuple(reversed(result_shape))

    def broadcast_to(self, target_shape):
        if self.shape == target_shape:
            return self
        new_data = [self.get(self.indices_from_flat_index(i, self.shape))
                    for i in range(self.size(target_shape))]
        return Array(target_shape, new_data)

    def indices_from_flat_index(self, flat_index, shape):
        indices = []
        stride = 1
        for s in reversed(shape):
            indices.append((flat_index // stride) % s)
            stride *= s
        return list(reversed(indices))

    def size(self, shape=None):
        if shape is None:
            shape = self.shape
        size = 1
        for s in shape:
            size *= s
        return size

    def __add__(self, other):
        if self.shape != other.shape:
            new_shape = self.broadcast_shape(other.shape)
            array1 = self.broadcast_to(new_shape)
            array2 = other.broadcast_to(new_shape)
        else:
            array1 = self
            array2 = other
        new_data = [a + b for a, b in zip(array1.data, array2.data)]
        return Array(array1.shape, new_data)

    def __sub__(self, other):
        if self.shape != other.shape:
            new_shape = self.broadcast_shape(other.shape)
            array1 = self.broadcast_to(new_shape)
            array2 = other.broadcast_to(new_shape)
        else:
            array1 = self
            array2 = other
        new_data = [a - b for a, b in zip(array1.data, array2.data)]
        return Array(array1.shape, new_data)

    def __truediv__(self, other):
        if self.shape != other.shape:
            new_shape = self.broadcast_shape(other.shape)
            array1 = self.broadcast_to(new_shape)
            array2 = other.broadcast_to(new_shape)
        else:
            array1 = self
            array2 = other
        new_data = [a / b for a, b in zip(array1.data, array2.data)]
        return Array(array1.shape, new_data)

    def __mul__(self, other):
        if isinstance(other, Array):
            if self.shape != other.shape:
                new_shape = self.broadcast_shape(other.shape)
                array1 = self.broadcast_to(new_shape)
                array2 = other.broadcast_to(new_shape)
            else:
                array1 = self
                array2 = other
            new_data = [a * b for a, b in zip(array1.data, array2.data)]
            return Array(array1.shape, new_data)
        elif isinstance(other, int) or isinstance(other, float):
            new_data = [a * other for a in self.data]
            return Array(self.shape, new_data)

    # def dot(self, other):
    #     if isinstance(other, Array):
    #         if len(self.shape) == 2 and len(other.shape) == 2 and self.shape[1] == other.shape[0]:
    #             result_data = []
    #             for i in range(self.shape[0]):
    #                 row = []
    #                 for j in range(other.shape[1]):
    #                     element = 0
    #                     for k in range(self.shape[1]):
    #                         element += self.get((i, k)) * other.get((k, j))
    #                     row.append(element)
    #                 result_data.extend(row)
    #             return Array((self.shape[0], other.shape[1]), result_data)
    #         else:
    #             raise ValueError("Incompatible shapes for dot product")
    #     else:
    #         raise TypeError("Unsupported type for dot product")

    def __pow__(self, exponent):
        new_data = [a ** exponent for a in self.data]
        return Array(self.shape, new_data)

    def __neg__(self):
        new_data = [-a for a in self.data]
        return Array(self.shape, new_data)

    def sum(self):
        return sum(self.data)

    def mean(self):
        return sum(self.data) / len(self.data)

    def product(self):
        result = 1
        for a in self.data:
            result *= a
        return result

    def min(self):
        return min(self.data)

    def max(self):
        return max(self.data)

    def sqrt(self):
        new_data = [a ** 0.5 for a in self.data]
        return Array(self.shape, new_data)

    def sin(self):
        import math
        new_data = [math.sin(a) for a in self.data]
        return Array(self.shape, new_data)

    def exp(self):
        import math
        new_data = [math.exp(a) for a in self.data]
        return Array(self.shape, new_data)

    def log(self):
        import math
        new_data = [math.log(a) for a in self.data]
        return Array(self.shape, new_data)

    # def dot(self, other):
    #     result = sum([a * b for a, b in zip(self.data, other.data)])
    #     return Array((1,), [result])
    def dot(self, other):
        if isinstance(other, Array):
            if len(self.shape) == 2 and len(other.shape) == 2:
                if self.shape[1] == other.shape[0]:
                    result_data = []
                    for i in range(self.shape[0]):
                        row = []
                        for j in range(other.shape[1]):
                            element = sum(a * b for a, b in zip(self.get_row(i), other.get_column(j)))
                            row.append(element)
                        result_data.extend(row)
                    return Array((self.shape[0], other.shape[1]), result_data)
                else:
                    raise ValueError("Incompatible shapes for dot product")
            elif len(self.shape) == 1 and len(other.shape) == 1:
                result_data = sum(a * b for a, b in zip(self.data, other.data))
                return result_data
            elif len(self.shape) == 1 and len(other.shape) == 2 and self.shape[0] == other.shape[0]:
                result_data = [sum(a * b for a, b in zip(self.data, other.get_column(j))) for j in range(other.shape[1])]
                return Array((other.shape[1],), result_data)
            elif len(self.shape) == 2 and len(other.shape) == 1 and self.shape[1] == other.shape[0]:
                result_data = [sum(a * b for a, b in zip(self.get_row(i), other.data)) for i in range(self.shape[0])]
                return Array((self.shape[0],), result_data)
            elif self.shape == other.shape:  # Handle element-wise multiplication
                result_data = [a * b for a, b in zip(self.data, other.data)]
                return Array(self.shape, result_data)
            else:
                raise ValueError("Incompatible shapes for dot product")
        else:
            raise TypeError("Unsupported type for dot product")

    def get_row(self, index):
        return [self.get((index, j)) for j in range(self.shape[1])]

    def get_column(self, index):
        return [self.get((i, index)) for i in range(self.shape[0])]

def main():
    array_shape1 = (2, 3)
    array_shape2 = (1, 3)

    array_data1 = [1, 2, 3, 4, 5, 6]
    array_data2 = [7, 8, 9]

    array1 = Array(array_shape1, array_data1)
    array2 = Array(array_shape2, array_data2)

    # Your Array operations
    addition_result = array1 + array2
    subtraction_result = array1 - array2
    multiplication_result = array1 * array2
    division_result = array1 / array2
    exponentiation_result = array1 ** 2
    negation_result = -array1
    sum_result = array1.sum()
    mean_result = array1.mean()
    product_result = array1.product()
    min_result = array1.min()
    max_result = array1.max()
    sqrt_result = array1.sqrt()
    sin_result = array1.sin()
    exp_result = array1.exp()
    log_result = array1.log()
    # dot_product_result = array1.dot(array2)

    # Convert Array instances to NumPy arrays
    np_array1 = np.array(array_data1).reshape(array_shape1)
    np_array2 = np.array(array_data2).reshape(array_shape2)

    # NumPy operations
    np_addition_result = np_array1 + np_array2
    np_subtraction_result = np_array1 - np_array2
    np_multiplication_result = np_array1 * np_array2
    np_division_result = np_array1 / np_array2
    np_exponentiation_result = np_array1 ** 2
    np_negation_result = -np_array1
    np_sum_result = np_array1.sum()
    np_mean_result = np_array1.mean()
    np_product_result = np_array1.prod()
    np_min_result = np_array1.min()
    np_max_result = np_array1.max()
    np_sqrt_result = np.sqrt(np_array1)
    np_sin_result = np.sin(np_array1)
    np_exp_result = np.exp(np_array1)
    np_log_result = np.log(np_array1)
    # np_dot_product_result = np.dot(np_array1, np_array2)

    # Compare results
    print("Addition (Array):", addition_result.data)
    print("Addition (NumPy):", np_addition_result)
    print("Arrays are equal:", np.array_equal(addition_result.data, np_addition_result))

    print("Subtraction (Array):", subtraction_result.data)
    print("Subtraction (NumPy):", np_subtraction_result)
    print("Arrays are equal:", np.array_equal(subtraction_result.data, np_subtraction_result))

    print("Multiplication (Array):", multiplication_result.data)
    print("Multiplication (NumPy):", np_multiplication_result)
    print("Arrays are equal:", np.array_equal(multiplication_result.data, np_multiplication_result))

    print("Division (Array):", division_result.data)
    print("Division (NumPy):", np_division_result)
    print("Arrays are equal:", np.array_equal(division_result.data, np_division_result))

    print("Exponentiation (Array):", exponentiation_result.data)
    print("Exponentiation (NumPy):", np_exponentiation_result)
    print("Arrays are equal:", np.array_equal(exponentiation_result.data, np_exponentiation_result))

    print("Negation (Array):", negation_result.data)
    print("Negation (NumPy):", np_negation_result)
    print("Arrays are equal:", np.array_equal(negation_result.data, np_negation_result))

    print("Sum (Array):", sum_result)
    print("Sum (NumPy):", np_sum_result)
    print("Values are equal:", sum_result == np_sum_result)
    
    print("Mean (Array):", mean_result)
    print("Mean (NumPy):", np_mean_result)
    print("Values are equal:", np.isclose(mean_result, np_mean_result))

    print("Product (Array):", product_result)
    print("Product (NumPy):", np_product_result)
    print("Values are equal:", np.isclose(product_result, np_product_result))

    print("Min (Array):", min_result)
    print("Min (NumPy):", np_min_result)
    print("Values are equal:", np.isclose(min_result, np_min_result))

    print("Max (Array):", max_result)
    print("Max (NumPy):", np_max_result)
    print("Values are equal:", np.isclose(max_result, np_max_result))

    print("Sqrt (Array):", sqrt_result.data)
    print("Sqrt (NumPy):", np_sqrt_result)
    print("Arrays are equal:", np.array_equal(sqrt_result.data, np_sqrt_result))

    print("Sin (Array):", sin_result.data)
    print("Sin (NumPy):", np_sin_result)
    print("Arrays are equal:", np.array_equal(sin_result.data, np_sin_result))

    print("Exp (Array):", exp_result.data)
    print("Exp (NumPy):", np_exp_result)
    print("Arrays are equal:", np.array_equal(exp_result.data, np_exp_result))

    print("Log (Array):", log_result.data)
    print("Log (NumPy):", np_log_result)
    print("Arrays are equal:", np.array_equal(log_result.data, np_log_result))
    
    array_shape1 = (2, 3)
    array_shape2 = (3, 2)
    array_shape3 = (2,)
    array_shape4 = (3,)
    array_shape5 = (3, 3)

    array_data1 = [1, 2, 3, 4, 5, 6]
    array_data2 = [7, 8, 9, 10, 11, 12]
    array_data3 = [2, 3]
    array_data4 = [4, 5, 6]
    array_data5 = [1, 2, 3, 4, 5, 6, 7, 8, 9]

    array1 = Array(array_shape1, array_data1)
    array2 = Array(array_shape2, array_data2)
    array3 = Array(array_shape3, array_data3)
    array4 = Array(array_shape4, array_data4)
    array5 = Array(array_shape5, array_data5)

    # Your Array dot operations
    dot_product1 = array1.dot(array2)
    # dot_product2 = array1.dot(array3)
    dot_product3 = array4.dot(array2)
    dot_product4 = array3.dot(array4)
    dot_product5 = array5.dot(array5)

    print("Dot Product 1 (Array):")
    print(dot_product1.data)
    print("Dot Product 2 (Array):")
    # print(dot_product2.data)
    print("Dot Product 3 (Array):")
    print(dot_product3.data)
    print("Dot Product 4 (Array):")
    # print(dot_product4.data)
    print("Dot Product 5 (Array):")
    print(dot_product5.data)

    # Convert Array instances to NumPy arrays
    np_array1 = np.array(array_data1).reshape(array_shape1)
    np_array2 = np.array(array_data2).reshape(array_shape2)
    np_array3 = np.array(array_data3).reshape(array_shape3)
    np_array4 = np.array(array_data4).reshape(array_shape4)
    np_array5 = np.array(array_data5).reshape(array_shape5)

    # NumPy dot operations
    np_dot_product1 = np.dot(np_array1, np_array2)
    # np_dot_product2 = np.dot(np_array1, np_array3)
    np_dot_product3 = np.dot(np_array4, np_array2)
    # np_dot_product4 = np.dot(np_array3, np_array4)
    np_dot_product5 = np.dot(np_array5, np_array5)

    print("Dot Product 1 (NumPy):")
    print(np_dot_product1)
    print("Dot Product 2 (NumPy):")
    # print(np_dot_product2)
    print("Dot Product 3 (NumPy):")
    print(np_dot_product3)
    print("Dot Product 4 (NumPy):")
    # print(np_dot_product4)
    print("Dot Product 5 (NumPy):")
    print(np_dot_product5)

    # Compare results
    print("Arrays are equal (Dot Product 1):", np.array_equal(dot_product1.data, np_dot_product1))
    # print("Arrays are equal (Dot Product 2):", np.array_equal(dot_product2.data, np_dot_product2))
    print("Arrays are equal (Dot Product 3):", np.array_equal(dot_product3.data, np_dot_product3))
    # print("Arrays are equal (Dot Product 4):", np.array_equal(dot_product4.data, np_dot_product4))
    print("Arrays are equal (Dot Product 5):", np.array_equal(dot_product5.data, np_dot_product5))



if __name__ == "__main__":
    main()

File: rust-NN/src/op2.rs
extern crate rust_math; // Ajouter la dépendance à la crate rust-math

use rust_math::math::exp; // Importer la fonction exp de la crate rust-math
use rust_math::math::log; // Importer la fonction log de la crate rust-math
use rust_math::math::sin; // Importer la fonction sin de la crate rust-math
use std::cmp;
use std::ops::{Add, Div, Mul, Sub};

#[derive(Clone)]
struct Array {
    shape: Vec<usize>,
    data: Vec<f64>,
}

impl Array {
    fn new(shape: Vec<usize>, data: Vec<f64>) -> Self {
        Self { shape, data }
    }

    fn get_index(&self, indices: &[usize]) -> usize {
        let mut flat_index = 0;
        let mut stride = 1;
        for i in (0..indices.len()).rev() {
            flat_index += indices[i] * stride;
            stride *= self.shape[i];
        }
        flat_index
    }

    fn get(&self, indices: &[usize]) -> f64 {
        let index = self.get_index(indices);
        self.data[index]
    }

    fn broadcast_shape(&self, other_shape: &[usize]) -> Vec<usize> {
        let mut result_shape = Vec::new();
        for (s1, s2) in self.shape.iter().rev().zip(other_shape.iter().rev()) {
            if *s1 == *s2 || *s1 == 1 || *s2 == 1 {
                result_shape.push(cmp::max(*s1, *s2));
            } else {
                panic!("Incompatible shapes for broadcasting");
            }
        }
        result_shape.reverse();
        result_shape
    }

    fn broadcast_to(&self, target_shape: &[usize]) -> Self {
        if self.shape == target_shape {
            return self.clone();
        }
        let new_data = (0..self.size(Some(target_shape)))
            .map(|i| {
                let indices = self.indices_from_flat_index(i, target_shape);
                self.get(&indices)
            })
            .collect();
        Self::new(target_shape.to_vec(), new_data)
    }

    fn indices_from_flat_index(&self, flat_index: usize, shape: &[usize]) -> Vec<usize> {
        let mut indices = Vec::new();
        let mut stride = 1;
        for &s in shape.iter().rev() {
            indices.push((flat_index / stride) % s);
            stride *= s;
        }
        indices.reverse();
        indices
    }

    fn size(&self, shape: Option<&[usize]>) -> usize {
        let target_shape = match shape {
            Some(s) => s,
            None => &self.shape,
        };
        target_shape.iter().product()
    }
}

impl Add<&Array> for &Array {
    type Output = Array;

    fn add(self, other: &Array) -> Array {
        let new_shape = self.broadcast_shape(&other.shape);
        let array1 = self.broadcast_to(&new_shape);
        let array2 = other.broadcast_to(&new_shape);
        let new_data: Vec<f64> = array1
            .data
            .iter()
            .zip(array2.data.iter())
            .map(|(a, b)| a + b)
            .collect();
        Array::new(array1.shape, new_data)
    }
}

impl Sub<&Array> for &Array {
    type Output = Array;

    fn sub(self, other: &Array) -> Array {
        let new_shape = self.broadcast_shape(&other.shape);
        let array1 = self.broadcast_to(&new_shape);
        let array2 = other.broadcast_to(&new_shape);
        let new_data: Vec<f64> = array1
            .data
            .iter()
            .zip(array2.data.iter())
            .map(|(a, b)| a - b)
            .collect();
        Array::new(array1.shape, new_data)
    }
}

impl Mul<&Array> for &Array {
    type Output = Array;

    fn mul(self, other: &Array) -> Array {
        let new_shape = self.broadcast_shape(&other.shape);
        let array1 = self.broadcast_to(&new_shape);
        let array2 = other.broadcast_to(&new_shape);
        let new_data: Vec<f64> = array1
            .data
            .iter()
            .zip(array2.data.iter())
            .map(|(a, b)| a * b)
            .collect();
        Array::new(array1.shape, new_data)
    }
}

impl Div<&Array> for &Array {
    type Output = Array;

    fn div(self, other: &Array) -> Array {
        let new_shape = self.broadcast_shape(&other.shape);
        let array1 = self.broadcast_to(&new_shape);
        let array2 = other.broadcast_to(&new_shape);
        let new_data: Vec<f64> = array1
            .data
            .iter()
            .zip(array2.data.iter())
            .map(|(a, b)| a / b)
            .collect();
        Array::new(array1.shape, new_data)
    }
}

impl Mul<f64> for &Array {
    type Output = Array;

    fn mul(self, scalar: f64) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| a * scalar).collect();
        Array::new(self.shape.clone(), new_data)
    }
}

impl Array {
    fn dot(&self, other: &Array) -> Array {
        if self.shape.len() == 2 && other.shape.len() == 2 {
            if self.shape[1] == other.shape[0] {
                let mut result_data = Vec::new();
                for i in 0..self.shape[0] {
                    let mut row = Vec::new();
                    for j in 0..other.shape[1] {
                        let element = self
                            .get_row(i)
                            .iter()
                            .zip(other.get_column(j).iter())
                            .map(|(&a, &b)| a * b)
                            .sum();
                        row.push(element);
                    }
                    result_data.extend(row);
                }
                Array::new(vec![self.shape[0], other.shape[1]], result_data)
            } else {
                panic!("Incompatible shapes for dot product");
            }
        } else {
            panic!("Both arrays must be 2D for dot product");
        }
    }

    fn pow(&self, exponent: f64) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| a.powf(exponent)).collect();
        Array::new(self.shape.clone(), new_data)
    }

    fn neg(&self) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| -a).collect();
        Array::new(self.shape.clone(), new_data)
    }

    fn sum(&self) -> f64 {
        self.data.iter().sum()
    }

    fn mean(&self) -> f64 {
        self.data.iter().sum::<f64>() / self.data.len() as f64
    }

    fn product(&self) -> f64 {
        self.data.iter().product()
    }

    fn min(&self) -> f64 {
        *self
            .data
            .iter()
            .min_by(|&&a, &&b| a.partial_cmp(&b).unwrap())
            .unwrap()
    }

    fn max(&self) -> f64 {
        *self
            .data
            .iter()
            .max_by(|&&a, &&b| a.partial_cmp(&b).unwrap())
            .unwrap()
    }

    fn sqrt(&self) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| a.sqrt()).collect();
        Array::new(self.shape.clone(), new_data)
    }

    fn sin(&self) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| sin(a)).collect();
        Array::new(self.shape.clone(), new_data)
    }

    fn exp(&self) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| exp(a)).collect();
        Array::new(self.shape.clone(), new_data)
    }

    fn log(&self) -> Array {
        let new_data: Vec<f64> = self.data.iter().map(|&a| log(a)).collect();
        Array::new(self.shape.clone(), new_data)
    }
}

fn main() {
    let array_shape1: Vec<usize> = vec![2, 3];
    let array_shape2: Vec<usize> = vec![1, 3];

    let array_data1: Vec<f64> = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0];
    let array_data2: Vec<f64> = vec![7.0, 8.0, 9.0];

    let array1 = Array::new(array_shape1.clone(), array_data1.clone());
    let array2 = Array::new(array_shape2.clone(), array_data2.clone());

    // Your Array operations
    let addition_result = &array1 + &array2;
    let subtraction_result = &array1 - &array2;
    let multiplication_result = &array1 * &array2;
    let division_result = &array1 / &array2;
    let exponentiation_result = array1.pow(2.0);
    let negation_result = array1.neg();
    let sum_result = array1.sum();
    let mean_result = array1.mean();
    let product_result = array1.product();
    let min_result = array1.min();
    let max_result = array1.max();
    let sqrt_result = array1.sqrt();
    let sin_result = array1.sin();
    let exp_result = array1.exp();
    let log_result = array1.log();
    let dot_product_result = array1.dot(&array2);

    // Compare results
    println!("Addition (Array): {:?}", addition_result.data);
    println!("Subtraction (Array): {:?}", subtraction_result.data);
    println!("Multiplication (Array): {:?}", multiplication_result.data);
    println!("Division (Array): {:?}", division_result.data);
    println!("Exponentiation (Array): {:?}", exponentiation_result.data);
    println!("Negation (Array): {:?}", negation_result.data);
    println!("Sum (Array): {:?}", sum_result);
    println!("Mean (Array): {:?}", mean_result);
    println!("Product (Array): {:?}", product_result);
    println!("Min (Array): {:?}", min_result);
    println!("Max (Array): {:?}", max_result);
    println!("Sqrt (Array): {:?}", sqrt_result.data);
    println!("Sin (Array): {:?}", sin_result.data);
    println!("Exp (Array): {:?}", exp_result.data);
    println!("Log (Array): {:?}", log_result.data);
    println!("Dot Product (Array): {:?}", dot_product_result.data);
}
